controllers/base/knowledgebase_controller.go

```diff
@@ -139,23 +128,78 @@ func (r *KnowledgeBaseReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		return ctrl.Result{}, nil
 	}
 
+	// Add a finalizer.Then, we can define some operations which should
+	// occur before the KnowledgeBase to be deleted.
```

Comment:
`before` or `after`

Reply:
`Before`. The finalizer is used to clean up the resource before it is actually deleted.